File: backend-gin/internal/lib/utils/concurency.go
package utils

import (
	"errors"
	"fmt"
	"sync"
)

func BatchProcess[T any, K any](
	batch []T, transform func(T) (K, error), maxWorkers int,
) ([]K, error) {
	var wg sync.WaitGroup
	results := make([]K, len(batch))
	errs := make([]error, len(batch))
	usedWorkers := make(chan struct{}, maxWorkers)

	for i := range batch {
		wg.Add(1)
		go func() {
			defer wg.Done()
			usedWorkers <- struct{}{}
			defer func() { <-usedWorkers }()
			processed, err := transform(batch[i])
			results[i] = processed
			errs[i] = err
		}()
	}

	wg.Wait()
	err := errors.Join(errs...)
	if err != nil {
		return results, fmt.Errorf("batch processing failed:\n%w", err)
	}

	return results, nil
}

File: backend-axum/src/lib_/utils/concurency.rs
use std::fmt::Display;
use std::sync::Arc;
use tokio::sync::Semaphore;
use tokio::task::{JoinError, JoinSet};

#[derive(Debug)]
pub enum BatchError<T> {
    Panic(JoinError),
    TransformFailed(Vec<(usize, T)>),
}

impl<T: Display> Display for BatchError<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            BatchError::Panic(e) => write!(f, "task panicked: {e}"),
            BatchError::TransformFailed(errs) => {
                writeln!(f, "batch processing failed:")?;
                for (i, e) in errs {
                    writeln!(f, "  [{i}]: {e}")?;
                }
                Ok(())
            }
        }
    }
}

impl<T: std::error::Error> std::error::Error for BatchError<T> {}

pub async fn batch_process<Input, Output, Error, Callback, Fut>(
    batch: Vec<Input>,
    transform: Callback,
    max_workers: usize,
) -> Result<Vec<Output>, BatchError<Error>>
where
    Input: Send + 'static,
    Output: Send + 'static,
    Error: Send + 'static,
    Callback: Fn(Input) -> Fut + Send + Sync + 'static,
    Fut: Future<Output = Result<Output, Error>> + Send,
{
    let length = batch.len();
    let transform = Arc::new(transform);
    let semaphore = Arc::new(Semaphore::new(max_workers));
    let mut join_set = JoinSet::new();

    for (i, item) in batch.into_iter().enumerate() {
        let semaphore = semaphore.clone();
        let transform = transform.clone();
        join_set.spawn(async move {
            let _worker_spot = semaphore
                .acquire()
                .await
                .expect("semaphaore was unexpectedly closed");
            let task_result = transform(item).await;
            (i, task_result)
        });
    }

    let mut results: Vec<Option<Output>> = (0..length).map(|_| None).collect();
    let mut errors: Vec<(usize, Error)> = Vec::new();

    while let Some(res) = join_set.join_next().await {
        match res {
            Ok((i, Ok(val))) => results[i] = Some(val),
            Ok((i, Err(e))) => errors.push((i, e)),
            Err(err) => return Err(BatchError::Panic(err)),
        }
    }

    if !errors.is_empty() {
        return Err(BatchError::TransformFailed(errors));
    }

    Ok(results.into_iter().flatten().collect())
}
